vm/compiler/src/ledger/mod.rs

```diff
@@ -315,12 +324,29 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
         Ok(())
     }
 
+    /// Appends the given transaction to the memory pool.
+    pub fn add_to_prover_puzzle_memory_pool(&mut self, prover_puzzle_solution: ProverPuzzleSolution<N>) -> Result<()> {
+        // Ensure the transaction does not already exist.
+        if self.prover_puzzle_memory_pool.contains(&prover_puzzle_solution) {
```

Comment:
once the `prover_puzzle_memory_pool` is an `IndexMap`, the `contains` will no longer be needed, as long as the value will be the same or it's an `IndexSet` (as `insert` will indicate if the key had already existed)